src/solver/tron.jl

```diff
@@ -0,0 +1,403 @@
+    nsteps += 1
+    project_step!(s, x, α * d, ℓ, u)
+    slope, qs = compute_Hs_slope_qs!(Hs, H, s, g)
+    if qs <= μ₀ * slope
+      search = false
+    else
+      α *= 0.5
+    end
+  end
+  if α < 1.0 && α < brkmin
+    α = brkmin
+    project_step!(s, x, α * d, ℓ, u)
+    slope, qs = compute_Hs_slope_qs!(Hs, H, s, g)
+  end
+
+  αd = α * d
+  project_step!(s, x, αd, ℓ, u)
+  project!(x, x .+ αd, ℓ, u)
+
+  return x, s
+end
+
+"""`α, s = cauchy(x, H, g, Δ, ℓ, u; μ₀ = 1e-2, μ₁ = 1.0, σ=10.0)`
+
+Computes a Cauchy step `s = P(x - α g) - x` for
+
+    min  q(s) = ¹/₂sᵀHs + gᵀs     s.t.    ‖s‖ ≦ μ₁Δ,  ℓ ≦ x + s ≦ u,
+
+with the sufficient decrease condition
+
+    q(s) ≦ μ₀sᵀg.
+"""
+function cauchy(x::Vector, H::Union{AbstractMatrix,AbstractLinearOperator}, g::Vector,
+                Δ::Real, α::Real, ℓ::Vector, u::Vector;
+                μ₀::Real = 1e-2, μ₁::Real = 1.0, σ::Real = 10.0)
+  # TODO: Use brkmin to care for g direction
+  _, _, brkmax = breakpoints(x, -g, ℓ, u)
+  n = length(x)
+  s = zeros(n)
+  Hs = zeros(n)
+
+  project_step!(s, x, -α * g, ℓ, u)
+
+  # Interpolate or extrapolate
+  s_norm = norm(s)
+  if s_norm > μ₁ * Δ
+    interp = true
+  else
+    slope, qs = compute_Hs_slope_qs!(Hs, H, s, g)
+    interp = qs >= μ₀ * slope
+  end
+
+  if interp
+    search = true
+    while search
+      α /= σ
+      project_step!(s, x, -α * g, ℓ, u)
+      s_norm = norm(s)
+      if s_norm <= μ₁ * Δ
+        slope, qs = compute_Hs_slope_qs!(Hs, H, s, g)
+        search = qs >= μ₀ * slope
+      end
+      # TODO: Correctly assess why this fails
+      if α < 1e-200
+        throw("α too small (qs = $qs, slope = $slope)")
+        #stalled = true
+        #status = "α too small"
+        #break
+      end
+    end
+  else
+    search = true
+    αs = α
+    while search && α <= brkmax
+      α *= σ
+      project_step!(s, x, -α * g, ℓ, u)
+      s_norm = norm(s)
+      if s_norm <= μ₁ * Δ
+        slope, qs = compute_Hs_slope_qs!(Hs, H, s, g)
+        if qs <= μ₀ * slope
+          αs = α
+        end
+      else
+        search = false
+      end
+    end
+    # Recover the last successful step
+    α = αs
+    s = project_step!(s, x, -α * g, ℓ, u)
+  end
+  return α, s
+end
+
+"""`projected_newton!(x, H, g, Δ, gctol, s, max_cgiter, ℓ, u)`
+
+Compute an approximate solution `d` for
+
+    min q(d) = ¹/₂dᵀHs + dᵀg    s.t.    ℓ ≦ x + d ≦ u,  ‖d‖ ≦ Δ
+
+starting from `s`.  The steps are computed using the conjugate gradient method
+projected on the active bounds.
+"""
+function projected_newton!(x::Vector, H::Union{AbstractMatrix,AbstractLinearOperator},
+                          g::Vector, Δ::Real, cgtol::Real, s::Vector,
+                          ℓ::Vector, u::Vector;
+                          max_cgiter::Int = max(50,length(x)))
+  n = length(x)
+  status = ""
+
+  Hs = H * s
+
+  # Projected Newton Step
+  exit_optimal = false
+  exit_pcg = false
+  exit_itmax = false
+  iters = 0
+  project!(x, x .+ s, ℓ, u)
```

Comment:
Here's another allocation. There would probably be no allocation if we wrote directly
```
x .= max.(ℓ, min.(x .+ s, u))
```
but that would be less general because it wouldn't be possible to supply user-defined projections.